profiles/common/modules/features/nexteuropa_newsroom/nexteuropa_newsroom.blocks.inc

```diff
@@ -0,0 +1,889 @@
+  $items = array();
+  if ($nid) {
+    $query = new EntityFieldQuery();
+    $query->entityCondition('entity_type', 'node')
+        ->fieldCondition($field_name, 'target_id', $nid)
+        ->execute();
+    $result = $query->execute();
+    if (isset($result['node'])) {
+      $ids = array_keys($result['node']);
+
+      foreach ($ids as $id) {
+        $items[$id] = _nexteuropa_newsroom_get_node_title($id);
+      }
+    }
+  }
+  return $items;
+}
+
+/**
+ * Get related content from node.
+ *
+ * @param string $field_name
+ *   Field name.
+ * @param object $node
+ *   Current node.
+ *
+ * @return array
+ *   Related nodes.
+ */
+function _nexteuropa_nexteuropa_get_related_content($field_name, $node) {
+  $items = array();
+  $field = isset($node->$field_name) ? $node->$field_name : NULL;
+  if ($field && isset($field[LANGUAGE_NONE])) {
+    foreach ($field[LANGUAGE_NONE] as $item) {
+      $items[$item['target_id']] = isset($item['entity']) ? $item['entity']->title : _nexteuropa_newsroom_get_node_title($item['target_id']);
+    }
+  }
+  return $items;
+}
+
+/**
+ * Helper function to get the node title without loading the whole node object.
+ */
+function _nexteuropa_newsroom_get_node_title($nid) {
+  $cache_key = 'newsroom:node_title:' . $nid;
+  $cache = cache_get($cache_key, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+  if ($cache) {
+    $node_title = $cache->data;
+  }
+  else {
+    $node_title = db_query('SELECT title_field_value FROM {field_data_title_field} WHERE entity_id = :nid', array(':nid' => $nid))->fetchField();
+    if (empty($node_title)) {
+      $node_title = db_query('SELECT title FROM {node} WHERE nid = :nid', array(':nid' => $nid))->fetchField();
+    }
+    cache_set($cache_key, $node_title, NEXTEUROPA_NEWSROOM_CACHE_TABLE, time() + 3600);
+  }
+  return $node_title;
+}
+
+/**
+ * Returns the Newsroom title and link depending by taxonomy terms.
+ *
+ * @return string
+ *   Subject.
+ */
+function _nexteuropa_newsroom_subject() {
+  return l(t('Newsroom'), NewsroomHelper::getNewsroomUrl('all'));
+}
+
+/**
+ * Returns a block content.
+ *
+ * Loops through newsroom type at top level and display the corresponding
+ * content and link for each newsroom type.
+ *
+ * @return string
+ *   Content.
+ */
+function _nexteuropa_nexteuropa_newsroom_contents() {
+  $block_settings = _nexteuropa_newsroom_get_summary_block_tid();
+
+  $topic_tids = array();
+  $topic_name = array();
+  // Sticky items (Centrally published).
+  $sticky = TRUE;
+  if (!drupal_is_front_page()) {
+    if (!$block_settings['display']) {
+      return '';
+    }
+    $sticky = FALSE;
+    if (!empty($block_settings['topic'])) {
+      foreach ($block_settings['topic'] as $topic) {
+        $topic_tids[] = $topic->tid;
+        $topic_name[] = $topic->name;
+      }
+    }
+  }
+
+  $newsroom_content = _nexteuropa_newsroom_content($block_settings['type'], FALSE, implode('+', $topic_tids), $sticky, $block_settings['topic'], $block_settings['topic_operator']);
```

Comment:
You should separate this block to smaller parts because the execution time could be quite a big one with those still uncached views and other manipulations what are there. Try to use view cache and if the manipulations are still that amount try to organise the code a bit more unit-based instead of these long 'generator' functions.